src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.cs

```diff
@@ -174,49 +176,92 @@ private static Task<HttpResponseMessage> InnerSendAsync(HttpRequestMessage reque
 
         private static async Task<HttpResponseMessage> SendWithAuthAsync(HttpRequestMessage request, Uri authUri, ICredentials credentials, bool preAuthenticate, bool isProxyAuth, bool doRequestAuth, HttpConnectionPool pool, CancellationToken cancellationToken)
         {
+            // If preauth is enabled and this isn't proxy auth, try to get a basic credential from the
+            // preauth credentials cache, and if successful, set an auth header for it onto the request.
+            // Currently we only support preauth for Basic.
+            bool performedBasicPreauth = false;
             if (preAuthenticate)
             {
-                NetworkCredential credential = credentials.GetCredential(authUri, BasicScheme);
+                Debug.Assert(pool.PreAuthCredentials != null);
+                NetworkCredential credential;
+                lock (pool.PreAuthCredentials) // TODO #28045: Get rid of this lock.
+                {
+                    // Just look for basic credentials.  If in the future we support preauth
+                    // for other schemes, this will need to search in order of precedence.
+                    Debug.Assert(pool.PreAuthCredentials.GetCredential(authUri, NegotiateScheme) == null);
+                    Debug.Assert(pool.PreAuthCredentials.GetCredential(authUri, NtlmScheme) == null);
+                    Debug.Assert(pool.PreAuthCredentials.GetCredential(authUri, DigestScheme) == null);
+                    credential = pool.PreAuthCredentials.GetCredential(authUri, BasicScheme);
```

Comment:
I'm a little unclear how actually saving off the ICredential and retrieving it here it useful.

It seems like all we're really doing here is remembering that we did successful Basic auth against this particular Uri previously.  The credentials should always be the same as if we just retrieved them again from the original Credential, right?

Reply:
> against this particular Uri previously

Not just this particular Uri... the credential cache does prefix matching.

Reply:
Ok, but you're adding specifically adding authUri to the credential cache below.  So that would only matter if authUri ends in a '/', is that right?  I mean, if I successfully authenticate against '/abc/foo.html' and then another request comes for '/abc/bar.html', that won't trigger preauth, right?  I'm just trying to understand the behavior here...

Reply:
Correct.

Reply:
Actually, I'm not sure that's correct... one sec...

Reply:
Relevant RFC: https://tools.ietf.org/html/rfc7617#section-2.2

Reply:
Yeah, not correct.  This test:
```C#
using System;
using System.Net;

class Program
{
    static void Main()
    {
        var cred = new NetworkCredential("hello", "world");
        var cc = new CredentialCache();
        cc.Add(new Uri("http://bing.com/something/hello.html"), "Basic", cred);

        string[] tests =
        {
            "http://microsoft.com/something/hello.html",
            "http://bing.com/something/hello.html",
            "http://bing.com/something/world.html",
            "http://bing.com/something/",
            "http://bing.com/",
            "http://bing.com/hello.html",
            "http://bing.com/world.html",
            "http://bing.com/somethingelse/",
            "http://bing.com/somethingelse/hello,html",
        };

        foreach (string s in tests)
        {
            Console.WriteLine(s.PadRight(50) + ": " + cc.GetCredential(new Uri(s), "Basic"));
        }
    }
}
```
produces:
```
http://microsoft.com/something/hello.html         :
http://bing.com/something/hello.html              : System.Net.NetworkCredential
http://bing.com/something/world.html              : System.Net.NetworkCredential
http://bing.com/something/                        : System.Net.NetworkCredential
http://bing.com/                                  :
http://bing.com/hello.html                        :
http://bing.com/world.html                        :
http://bing.com/somethingelse/                    : System.Net.NetworkCredential
http://bing.com/somethingelse/hello,html          : System.Net.NetworkCredential
```

Reply:
(Seems like maybe CredentialCache has a bug and should be including the '/' in the test... those last two "somethingelse" entries look wrong to me.)

Reply:
BTW, there are similar (but different of course) rules for digest: https://tools.ietf.org/html/rfc7616#section-3.6, https://tools.ietf.org/html/rfc7616#section-3.3 (for definition of "protection space").  

Of course for digest, we need to remember some state from previous challenges in order to preauthenticate.

Reply:
yeah, those results look wacky...  but your theory re '/' at least makes them look somewhat sane.

Reply:
@davidsh, is this an issue with CredentialCache?  It seems like in the line here:
https://github.com/dotnet/corefx/blob/02c3892380a392458a4d1f41c53ce2048cb6681d/src/System.Net.Primitives/src/System/Net/CredentialCache.cs#L524
`prefixLen` should maybe be `prefixLen + 1`?

Reply:
>@davidsh, is this an issue with CredentialCache? 

I'm not sure. We haven't had any bug reports previously about this.

cc: @Tratcher comments?

Reply:
Agreed, the somethingelse match looks wrong. That said I haven't seen many users that scope by path, scoping by domain is the primary use case.